File: sdk/go/pty.go
package sdk

import (
	"context"
	"encoding/json"
	"errors"
	"io"
	"net/url"
	"sync"
	"time"

	"github.com/coder/websocket"

	"minisandbox/pkg/protocol"
)

// PTYRequest 是打开交互终端的启动请求。
type PTYRequest = protocol.PTYStartRequest

// PTYTerminal 是会话唯一终态结果。
type PTYTerminal = protocol.PTYServerEvent

// PTYOutputChunk 是终端输出块；PTY 天生合并 stdout 与 stderr。
type PTYOutputChunk []byte

// OpenPTY 在当前 sandbox 中打开一个交互式 PTY。
//
// 连接使用 minisandbox.pty.v1 子协议；发送 start 后等待 started 事件
// 才返回。一条连接对应一个会话；Close 即取消，进程组按服务端语义终止。
func (s *Sandbox) OpenPTY(ctx context.Context, request PTYRequest) (*PTYConnection, error) {
	if err := request.Validate(); err != nil {
		return nil, err
	}
	connection, _, err := websocket.Dial(
		ctx,
		s.client.baseURL+"/v1/sandboxes/"+url.PathEscape(s.id)+"/pty",
		&websocket.DialOptions{
			Subprotocols: []string{protocol.PTYSubprotocol},
		},
	)
	if err != nil {
		return nil, err
	}
	if connection.Subprotocol() != protocol.PTYSubprotocol {
		_ = connection.Close(websocket.StatusProtocolError, "subprotocol mismatch")
		return nil, errors.New("minisandbox: PTY subprotocol mismatch")
	}
	encoded, err := json.Marshal(request)
	if err != nil {
		_ = connection.Close(websocket.StatusInternalError, "encode start")
		return nil, err
	}
	writeContext, cancel := context.WithTimeout(ctx, 15*time.Second)
	defer cancel()
	if err := connection.Write(writeContext, websocket.MessageText, encoded); err != nil {
		_ = connection.Close(websocket.StatusInternalError, "write start")
		return nil, err
	}

	pty := &PTYConnection{
		connection: connection,
		output:     make(chan PTYOutputChunk, 64),
		terminal:   make(chan PTYTerminal, 1),
	}
	// 首个事件必须是 started；读取循环在后台持续消费。
	first, firstBinary, err := pty.readFrame(ctx)
	if err != nil {
		_ = connection.Close(websocket.StatusInternalError, "read started")
		return nil, err
	}
	if firstBinary {
		_ = connection.Close(websocket.StatusProtocolError, "expected started")
		return nil, errors.New("minisandbox: PTY first message must be started")
	}
	var started protocol.PTYServerEvent
	if err := json.Unmarshal(first, &started); err != nil || started.Type != protocol.PTYServerEventStarted {
		_ = connection.Close(websocket.StatusProtocolError, "invalid started")
		return nil, errors.New("minisandbox: PTY started event is invalid")
	}
	go pty.pump(ctx)
	return pty, nil
}

// PTYConnection 是一个已启动的交互式终端会话。
//
// Write 与 Resize 可并发调用；Output 与 Terminal 各自只投递一次序列。
type PTYConnection struct {
	connection *websocket.Conn

	output   chan PTYOutputChunk
	terminal chan PTYTerminal

	writeMu sync.Mutex
	closeMu sync.Mutex
	closed  bool
}

// Output 返回终端输出块通道；会话结束后通道关闭。
func (p *PTYConnection) Output() <-chan PTYOutputChunk { return p.output }

// Terminal 返回唯一终态事件通道。
func (p *PTYConnection) Terminal() <-chan PTYTerminal { return p.terminal }

// Write 向终端 stdin 写入字节；Ctrl-C 等控制字符按终端语义传递。
func (p *PTYConnection) Write(ctx context.Context, chunk []byte) error {
	p.writeMu.Lock()
	defer p.writeMu.Unlock()
	return p.connection.Write(ctx, websocket.MessageBinary, chunk)
}

// Resize 调整终端窗口大小。
func (p *PTYConnection) Resize(ctx context.Context, cols, rows uint16) error {
	encoded, err := json.Marshal(protocol.PTYResizeRequest{Type: protocol.PTYMessageTypeResize, Cols: cols, Rows: rows})
	if err != nil {
		return err
	}
	p.writeMu.Lock()
	defer p.writeMu.Unlock()
	return p.connection.Write(ctx, websocket.MessageText, encoded)
}

// Close 关闭会话；等价于取消，服务端终止整个进程组。
func (p *PTYConnection) Close() error {
	p.closeMu.Lock()
	defer p.closeMu.Unlock()
	if p.closed {
		return nil
	}
	p.closed = true
	return p.connection.Close(websocket.StatusNormalClosure, "sdk closed")
}

// pump 在后台消费服务端帧并分发到对应通道。
func (p *PTYConnection) pump(ctx context.Context) {
	defer close(p.output)
	for {
		payload, binary, err := p.readFrame(ctx)
		if err != nil {
			return
		}
		if binary {
			select {
			case p.output <- PTYOutputChunk(payload):
			case <-ctx.Done():
				return
			}
			continue
		}
		var event protocol.PTYServerEvent
		if json.Unmarshal(payload, &event) != nil {
			continue
		}
		if event.Type == protocol.PTYServerEventTerminal {
			p.terminal <- event
			close(p.terminal)
			_ = p.Close()
			return
		}
	}
}

// readFrame 读取一帧并完整返回其负载。
func (p *PTYConnection) readFrame(ctx context.Context) ([]byte, bool, error) {
	messageType, reader, err := p.connection.Reader(ctx)
	if err != nil {
		return nil, false, err
	}
	payload, err := io.ReadAll(reader)
	if err != nil {
		return nil, false, err
	}
	return payload, messageType == websocket.MessageBinary, nil
}

File: sdk/typescript/src/pty.ts
/**
 * 交互式 PTY：一条 WebSocket 连接对应一个会话。
 *
 * 仅支持 Node.js 22+ 的原生 WebSocket；帧语义与公共协议一致。
 */
import { sandboxBasePath, Transport } from "./transport.js";
import { PTYStartRequest, PTYTerminalEvent } from "./wire.js";

/** PTYConnection 是一个已启动的交互式终端会话。 */
export class PTYConnection {
  private readonly socket: WebSocket;
  private readonly pending: Array<{ resolve: (chunk: Uint8Array) => void; reject: (error: Error) => void }> = [];
  private readonly queue: Uint8Array[] = [];
  private terminalEvent?: PTYTerminalEvent;
  private terminalWaiters: Array<{
    resolve: (event: PTYTerminalEvent) => void;
    reject: (error: Error) => void;
  }> = [];
  private failure?: Error;

  private constructor(socket: WebSocket) {
    this.socket = socket;
    socket.binaryType = "arraybuffer";
    socket.onmessage = (event) => this.handleMessage(event.data);
    socket.onclose = () => this.handleClose();
    socket.onerror = () => {
      this.failure ??= new Error("minisandbox: PTY connection error");
      this.drainPending();
    };
  }

  /** 打开一个 PTY 会话并等待 started 事件。 */
  static open(transport: Transport, sandboxId: string, request: PTYStartRequest): Promise<PTYConnection> {
    return new Promise((resolve, reject) => {
      const url = `${transport.baseUrl.replace(/^http/, "ws")}${sandboxBasePath(sandboxId)}/pty`;
      const socket = new WebSocket(url, "minisandbox.pty.v1");
      socket.binaryType = "arraybuffer";
      const fail = (message: string) => {
        reject(new Error(message));
        try {
          socket.close();
        } catch {
          // 关闭失败时保持静默；promise 已经拒绝。
        }
      };
      socket.onopen = () => {
        socket.send(
          JSON.stringify({
            type: "start",
            argv: request.argv,
            ...(request.cwd !== undefined ? { cwd: request.cwd } : {}),
            ...(request.env !== undefined ? { env: request.env } : {}),
            cols: request.cols,
            rows: request.rows,
            ...(request.timeoutSeconds !== undefined ? { timeout_seconds: request.timeoutSeconds } : {}),
          }),
        );
      };
      socket.onerror = () => fail("minisandbox: PTY connection failed");
      socket.onclose = () => fail("minisandbox: PTY connection closed before start");
      socket.onmessage = (event) => {
        const connection = new PTYConnection(socket);
        socket.onmessage = (message) => connection.handleMessage(message.data);
        socket.onclose = () => connection.handleClose();
        socket.onerror = () => connection.handleError();
        const payload = typeof event.data === "string" ? event.data : "";
        let parsed: { type?: string };
        try {
          parsed = JSON.parse(payload);
        } catch {
          fail("minisandbox: PTY started event is invalid");
          return;
        }
        if (parsed.type !== "started") {
          fail(`minisandbox: PTY first message must be started, got ${parsed.type}`);
          return;
        }
        resolve(connection);
      };
    });
  }

  /** 向终端 stdin 写入字节。 */
  write(chunk: Uint8Array | string): void {
    const payload = typeof chunk === "string" ? chunk : chunk.buffer;
    this.socket.send(payload as BufferSource);
  }

  /** 调整终端窗口大小。 */
  resize(cols: number, rows: number): void {
    this.socket.send(JSON.stringify({ type: "resize", cols, rows }));
  }

  /** 关闭会话；等价于取消。 */
  close(): void {
    this.socket.close();
  }

  /** 逐块读取终端输出；PTY 天生合并 stdout 与 stderr。 */
  readOutput(): Promise<Uint8Array> {
    const buffered = this.queue.shift();
    if (buffered !== undefined) {
      return Promise.resolve(buffered);
    }
    if (this.failure !== undefined) {
      return Promise.reject(this.failure);
    }
    if (this.terminalEvent !== undefined || this.socket.readyState === WebSocket.CLOSED) {
      return Promise.reject(new Error("minisandbox: PTY output ended"));
    }
    return new Promise((resolve, reject) => {
      this.pending.push({ resolve, reject });
    });
  }

  /** 等待唯一终态事件。 */
  waitTerminal(): Promise<PTYTerminalEvent> {
    if (this.terminalEvent !== undefined) {
      return Promise.resolve(this.terminalEvent);
    }
    if (this.failure !== undefined) {
      return Promise.reject(this.failure);
    }
    return new Promise((resolve, reject) => {
      this.terminalWaiters.push({ resolve, reject });
    });
  }

  private handleMessage(data: unknown): void {
    if (typeof data === "string") {
      let parsed: { type?: string; exit_code?: number; duration_ms?: number; error_code?: string; message?: string };
      try {
        parsed = JSON.parse(data);
      } catch {
        return;
      }
      if (parsed.type === "terminal") {
        this.terminalEvent = {
          type: "terminal",
          exitCode: parsed.exit_code,
          durationMs: parsed.duration_ms,
          errorCode: parsed.error_code,
          message: parsed.message,
        };
        for (const waiter of this.terminalWaiters.splice(0)) {
          waiter.resolve(this.terminalEvent);
        }
        this.drainPending();
        this.close();
      }
      return;
    }
    const chunk = data instanceof ArrayBuffer ? new Uint8Array(data) : new Uint8Array(0);
    const waiter = this.pending.shift();
    if (waiter !== undefined) {
      waiter.resolve(chunk);
      return;
    }
    this.queue.push(chunk);
  }

  private handleClose(): void {
    if (this.terminalEvent === undefined) {
      this.failure ??= new Error("minisandbox: PTY connection closed");
    }
    this.drainPending();
    this.drainTerminalWaiters();
  }

  private handleError(): void {
    this.failure ??= new Error("minisandbox: PTY connection error");
    this.drainPending();
    this.drainTerminalWaiters();
  }

  private drainPending(): void {
    for (const waiter of this.pending.splice(0)) {
      waiter.reject(this.failure ?? new Error("minisandbox: PTY output ended"));
    }
  }

  private drainTerminalWaiters(): void {
    if (this.failure === undefined) {
      return;
    }
    for (const waiter of this.terminalWaiters.splice(0)) {
      waiter.reject(this.failure);
    }
  }
}

/** 打开一个交互式 PTY 会话。 */
export function openPTY(transport: Transport, sandboxId: string, request: PTYStartRequest): Promise<PTYConnection> {
  return PTYConnection.open(transport, sandboxId, request);
}
